src/rand.c

```diff
@@ -0,0 +1,81 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdint.h>
+#include "rand.h"
+#include "tinymt64.h"
+
+// TODO Add support for cryptographically secure PRNG
+// TODO Add Windows CNG
+
+static tinymt64_t fast_rand;
+
+#if HAVE_GETRANDOM_SYSCALL
+// -------------------
+// Modern linux kernel
+// -------------------
+// Use getrandom syscall which has some advantages over /dev/urandom
+
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <linux/random.h>
+
+void jq_rand_init() {
+  uint64_t seed;
+  if (syscall(SYS_getrandom, &seed, sizeof(seed), 0) != sizeof(seed)) {
+    fprintf(stderr, "error: could not get data from random pool.\n");
+    exit(2);
+  }
+  tinymt64_init(&fast_rand, seed);
+}
+
+#elif HAVE__DEV_URANDOM
+// -------------------
+// Unix machine
+// -------------------
+// Use /dev/urandom as fallback
+
+void jq_rand_init() {
+  uint64_t seed;
+  FILE *devur = fopen("/dev/urandom", "r");
+  if (fread(&seed, sizeof(seed), 1, devur) != 1) {
+    fprintf(stderr, "error: could not get data from /dev/urandom.\n");
+    exit(2);
+  }
+  if (fclose(devur)) {
+    fprintf(stderr, "error: could not close /dev/urandom.\n");
+    exit(2);
+  }
+  tinymt64_init(&fast_rand, seed);
+}
+
+#else
+// --------
+// Fallback
+// --------
+// Always use MT, seeded with time if no better randomness source exists
+
+#include <time.h>
+
+void jq_rand_init() {
+  tinymt64_init(&fast_rand, time(NULL));
+}
+
+#endif
+
+// Common functions independent of randomness source
+void jq_rand_init_seed(uint64_t seed) {
+  tinymt64_init(&fast_rand, seed);
+}
+
+uint64_t jq_rand_int(uint64_t max_val) {
+  uint64_t thresh = -1 - ((-1 % max_val) + 1) % max_val;
+  uint64_t result;
+  while (thresh < (result = tinymt64_generate_uint64(&fast_rand))); // Efficient rejection sample
+  return result % max_val;
+}
+
+double jq_rand_double() {
+  return tinymt64_generate_double01(&fast_rand);
```

Comment:
Why wouldn't we *always* read from `/dev/urandom` (or `getentropy(2)` or `getrandom(2)` or CNG)?

I'm not keen on having our own PRNG.  That way lies trouble.

Reply:
Also, what's `tinymt64_generate_double01()`?

Reply:
As I've mentioned previously, my motivation isn't cryptographic security, but often reproducability, which is why I used a PRNG and allowed specifying a seed. I agree that out own PRNG spell trouble, which is why I'm using code from the authors of Mersenne Twister as a submodule. tinymy64_generate_double01 is the name of their function that returns the next double in [0, 1).